File: src/lib/validateSubmission.ts
export type ValidationDecision = "approved" | "not_approved" | "on_hold";

export interface ValidationResult {
  decision: ValidationDecision;
  reasoning: string;
}

/**
 * Calls the OpenAI API to determine the validation status of a Kobo submission.
 * Returns the AI's decision and reasoning, or null on failure.
 */
export async function callValidationAI(
  apiKey: string,
  submission: Record<string, unknown>,
  instructions: string,
  options: { approved: string; notApproved: string; onHold: string }
): Promise<ValidationResult | null> {
  const systemPrompt = [
    "You are a submission reviewer. Review the following form submission and decide on its validation status.",
    "",
    instructions ? `Overall context: ${instructions}` : "",
    "",
    "Criteria for each status:",
    `- Approved: ${options.approved || "The submission meets all requirements."}`,
    `- Not Approved: ${options.notApproved || "The submission does not meet requirements."}`,
    `- On Hold: ${options.onHold || "The submission needs further review."}`,
    "",
    'Respond with valid JSON only — no markdown fences, no extra text:',
    '{"decision":"approved"|"not_approved"|"on_hold","reasoning":"<explanation>"}',
  ]
    .filter((l) => l !== undefined)
    .join("\n");

  try {
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: "gpt-4o-mini",
        messages: [
          { role: "system", content: systemPrompt },
          {
            role: "user",
            content: `Submission data:\n${JSON.stringify(submission, null, 2)}`,
          },
        ],
        max_tokens: 512,
      }),
    });

    if (!res.ok) {
      console.error(`[validate] OpenAI error ${res.status}`);
      return null;
    }

    const data = await res.json<{ choices: Array<{ message: { content: string } }> }>();
    const raw = data.choices?.[0]?.message?.content?.trim() ?? "";

    // Strip markdown code fences if the model returns them
    const cleaned = raw.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "").trim();

    let parsed: unknown;
    try {
      parsed = JSON.parse(cleaned);
    } catch {
      console.error(`[validate] Failed to parse AI response: ${raw.slice(0, 200)}`);
      return null;
    }

    if (
      typeof parsed !== "object" ||
      parsed === null ||
      !["approved", "not_approved", "on_hold"].includes((parsed as Record<string, unknown>).decision as string)
    ) {
      console.error(`[validate] Unexpected AI response shape: ${cleaned.slice(0, 200)}`);
      return null;
    }

    const result = parsed as { decision: string; reasoning: string };
    return {
      decision: result.decision as ValidationDecision,
      reasoning: typeof result.reasoning === "string" ? result.reasoning : "",
    };
  } catch (err) {
    console.error("[validate] callValidationAI error:", err);
    return null;
  }
}

File: app/lib/validate_submission.py
"""
AI-based Kobo submission validation.
Port of src/lib/validateSubmission.ts.
"""

import json
import logging
import re

import httpx

logger = logging.getLogger(__name__)

VALID_DECISIONS = {'approved', 'not_approved', 'on_hold'}


def call_validation_ai(
    api_key: str,
    submission: dict,
    instructions: str,
    options: dict[str, str],
) -> dict | None:
    """
    Calls OpenAI to determine validation status of a submission.
    Returns {'decision': str, 'reasoning': str} or None on failure.
    """
    system_prompt = '\n'.join([
        'You are a submission reviewer. Review the following form submission and decide on its validation status.',
        '',
        f'Overall context: {instructions}' if instructions else '',
        '',
        'Criteria for each status:',
        f'- Approved: {options.get("approved") or "The submission meets all requirements."}',
        f'- Not Approved: {options.get("notApproved") or "The submission does not meet requirements."}',
        f'- On Hold: {options.get("onHold") or "The submission needs further review."}',
        '',
        'Respond with valid JSON only — no markdown fences, no extra text:',
        '{"decision":"approved"|"not_approved"|"on_hold","reasoning":"<explanation>"}',
    ])

    try:
        resp = httpx.post(
            'https://api.openai.com/v1/chat/completions',
            headers={'Authorization': f'Bearer {api_key}', 'Content-Type': 'application/json'},
            json={
                'model': 'gpt-4o-mini',
                'messages': [
                    {'role': 'system', 'content': system_prompt},
                    {'role': 'user', 'content': f'Submission data:\n{json.dumps(submission, indent=2)}'},
                ],
                'max_tokens': 512,
            },
            timeout=30,
        )
        if not resp.is_success:
            logger.error('[validate] OpenAI error %s', resp.status_code)
            return None

        raw = (resp.json().get('choices') or [{}])[0].get('message', {}).get('content', '').strip()
        # Strip markdown code fences if present
        cleaned = re.sub(r'^```(?:json)?\s*', '', raw, flags=re.IGNORECASE)
        cleaned = re.sub(r'\s*```$', '', cleaned).strip()

        parsed = json.loads(cleaned)
        decision = parsed.get('decision', '')
        if decision not in VALID_DECISIONS:
            logger.error('[validate] Unexpected AI response shape: %s', cleaned[:200])
            return None

        return {
            'decision': decision,
            'reasoning': str(parsed.get('reasoning', '')),
        }
    except json.JSONDecodeError:
        logger.error('[validate] Failed to parse AI response')
        return None
    except Exception as exc:
        logger.error('[validate] callValidationAI error: %s', exc)
        return None
